### ISSUE TITLE
Dealing with potential frame shape mismatch for batch APIs

### ISSUE THREAD

**Issue by NicolasHug** [pytorch/torchcodec#433]
Pre-read: the current state of output tensor allocation as described in the comment below:

https://github.com/pytorch/torchcodec/blob/84cef500b5865c44fe5c799a2f468c2067d31220/src/torchcodec/decoders/_core/VideoDecoder.h#L430-L473

-----

This issue is about addressing that specific part of the comment:

https://github.com/pytorch/torchcodec/blob/84cef500b5865c44fe5c799a2f468c2067d31220/src/torchcodec/decoders/_core/VideoDecoder.h#L464-L474

Basically, the idea is that instead of allocating batch output tensors based on H and W from the metadata, we could allocate them only *after* we have decoded the first frame, and rely on the H and W from that decoded frame, which is the source of truth for dimensions.

**Comment by NicolasHug** [comment]
Yeah. I don't think we'll do that. At least not until we see a very very good reason to.
